ethereum/slogging.py

```diff
@@ -138,11 +85,228 @@ def pop_records(self):
+        #getLogger()
+        #self.setLevel(getattr(logging, level_name.upper()))
 
 
+class RootLogger(ethLogger):
+    """
+    A root logger is not that different to any other logger, except that
+    it must have a logging level and there is only one instance of it in
+    the hierarchy.
+    """
+    def __init__(self, level, log_json=False):
+        """
+        Initialize the logger with the name "root".
+        """
+        ethLogger.__init__(self, "root", level)
+        self.log_json = False
+
+class ethManager(logging.Manager):
+    def __init__(self, rootnode):
+        self.loggerClass = ethLogger
+        super(ethManager, self).__init__(rootnode)
+
+    def getLogger(self, name):
+        logging.setLoggerClass(ethLogger)
+        return super(ethManager, self).getLogger(name)
+
+rootLogger = RootLogger(DEFAULT_LOGLEVEL)
+ethLogger.root = rootLogger
+ethLogger.manager = ethManager(ethLogger.root)
+
+def help_make_kws(kws, name, msg, log_json=False):
+    new_kws = {}
+    message = ""
+
+    if log_json == True:
+        message = {}
+        for i in kws:
+            if i not in formatter_list:  # levelno, asctime......
+                print i, "=", kws[i]
+                #s = " {i}={kwsi}".format(i=i, kwsi=str(kws[i]))
+                #message += s
+                #message[i] = repr(kws[i]) if type(kws[i]).__name__=='instance'   else kws[i]
+                #if type(kws[i]) == (types.IntType or types.LongType or types.FloatType or types.ComplexType or types.DictionaryType):
+                #    message[i] = kws[i] # for json dumps
+                #else:
+                #    message[i] = repr(kws[i]) # for json dumps
+                kws_i_d = getattr(kws[i], "__dict__", False)
+                if kws_i_d != False:
+                    #message[i] = "{} {}".format(repr(kws[i]), kws[i].__dict__)
+                    message[i] = "{}".format(kws[i].__dict__)
+                else:
+                    message[i] = kws[i]
+            else:
+                new_kws[i] = kws[i]
+    else:
+        message = ""
+        for i in kws:
+            if i not in formatter_list:
+                print i, "=", kws[i]
+                s = " {i}={kwsi}".format(i=i, kwsi=str(kws[i]))
+                message += s
+            else:
+                new_kws[i] = kws[i]
+
+    if log_json == True:
+        if name != rootLogger.name:
+            message1 = {'event': "{}.{}".format(name, msg)}
+        else:
+            message1 = {'event': "{}".format(msg)}
+        message.update(message1)
+        msg = json.dumps(message)
+    else:
+        msg += message
+
+    return new_kws, msg
+
+# this function extends standart logging module
+def _trace(self, msg, *args, **kwargs):
+    # Yes, logger takes its '*args' as 'args'.
+
+    if self.isEnabledFor(TRACE):
+        new_kws, new_message = help_make_kws(kwargs, self.name, msg, self.log_json if hasattr(self, "log_json") else False )
```

Comment:
Can be simplified to `... = help_make_kws(kwargs, self.name, msg, getattr(self, 'log_json', False))`